kafka/producer/base.py

```diff
@@ -71,12 +84,48 @@ def _send_upstream(queue, client, codec, batch_time, batch_size,
                                  messages)
             reqs.append(req)
 
+        if not reqs:
+            continue
+
+        reqs_to_retry, error_type = [], None
         try:
             client.send_produce_request(reqs,
                                         acks=req_acks,
                                         timeout=ack_timeout)
-        except Exception:
-            log.exception("Unable to send message")
+
+        except FailedPayloadsError as ex:
+            error_type = FailedPayloadsError
+            reqs_to_retry = ex.failed_payloads
+
+        except RequestTimedOutError:
+            error_type = RequestTimedOutError
+            if retry_options.retry_on_timeouts:
+                reqs_to_retry = reqs
+
+        except Exception as ex:
+            error_type = type(ex)
+            if type(ex) in RETRY_ERROR_TYPES:
+                reqs_to_retry = reqs
+
+        finally:
+            reqs = []
+
+        if not reqs_to_retry or retry_options.limit == 0:
```

Comment:
I dont think we want a retry_options check here to prevent us from refreshing metadata on a refresh_error